### PR TITLE
Rewrite PartialResultStream

### LINKED ISSUE #180: runStream returning incomplete and out of order rows
#### Environment details

  - OS: Ubuntu 16.04.4 LTS
  - Node.js version: 6.14.1
  - npm version: 5.8.0
  - `@google-cloud/spanner` version: 1.4.1

#### Steps to reproduce

I have a table with this structure:
```
CREATE TABLE accounts (
    recordId STRING(36) NOT NULL,
    account_CID STRING(255),
    account_created_on STRING(1024),
    account_created_on_unixTimestamp FLOAT64,
    account_dashboard_tabs STRING(255),
    account_ok_tabs STRING(255),
    account_summary_content_tabs STRING(255),
    account_to_creator STRING(36),
    account_to_customer_admin STRING(36),
    account_view_tabs STRING(255),
    Address STRING(1024),
    Address1 STRING(1024),
    Address1_city STRING(255),
    Address1_postalcode STRING(255),
    Address1_provinceterritory STRING(255),
    Address_city STRING(255),
    Address_postalcode STRING(255),
    Address_provinceterritory STRING(255),
    current_document_number FLOAT64,
    customer_content_tabs STRING(255),
    Email STRING(255),
    fund_to_account STRING(36),
    FundsTab_AccountSubNavigation STRING(255),
    Number FLOAT64,
    Number1 FLOAT64,
    Number_of_Payments FLOAT64,
    payment_plan_to_account STRING(36),
    Phone_Number STRING(1024),
    Phone_Number1 STRING(1024),
    requested_payment_plan_to_account STRING(36),
    RequestRelationshipCID STRING(255),
    RequestRelationshipContentTab STRING(255),
    Selection_Field___Static STRING(MAX),
    Short_Text STRING(255),
    Short_Text1 STRING(255),
    Short_Text4 STRING(255),
    upload_customer_csv STRING(1024),
    upload_customer_csv_name STRING(255),
    user_to_accounts STRING(36),
) PRIMARY KEY (recordId)
```
When I run the following query:
```
SELECT `root`.`account_created_on` as `field0`, `root`.`Short_Text` as `field1`, `root`.`Short_Text1` as `field2`, `root`.`account_CID` as `field3`, `root`.`recordId` as `recordId` FROM `accounts` AS root
```
My table has 17349  records

Occasionally (1 in 5) when I run this query using:
```js
database.runStream(query)
	.on('error', function (err) {
		console.error(err);
		return reject(err);
	})
	.on('data', function (row) {
		let rowJSON = row.toJSON();
		if (!rowJSON.recordId || rowJSON.recordId.match(/^\d+$/)) {
			console.log('Worker '+process.pid+' Row: '+JSON.stringify(rowJSON));
			console.log(row);
		}
		records[rowJSON.recordId] = rowJSON;
	})
	.on('end', function () {
		return resolve(records);
	});
```
I will get rows that are missing the recordId and the values will be the wrong labels for example field1 will contain the data for field2

### PR DESCRIPTION
**Pull request by stephenplusplus**
Fixes #180 and #198 by reworking the way we process streamed rows from PartialResultStream on Spanner. The high-level difference is this

1) use RowBuilder for all streams. We had two code paths previously
2) do not block on returning until we have a non-chunked row. We can hold the final value of a chunked segment and begin returning other data
3) handle partial segments. Occasionally we will get a partial row and shouldn't try to send that to the user.

#### To Dos

- [x] Unit tests

### CODE REVIEW

**Comment by stephenplusplus** [comment]
@crwilcox I made some changes, but my tests still pass. If this looks okay to you, I'll move onto unit tests.

File: src/partial-result-stream.js [hunk 1]
```diff
@@ -155,41 +134,4 @@ function partialResultStream(requestFn, options) {
...
-      row.values = valueSet;
-      return partialResultStream.formatRow_(metadata, row);
-    });
-  }
-
-  var formattedRow = row.values.map(function(value, index) {
-    var field = fields[index];
-    return {
-      name: field.name,
-      value: codec.decode(value, field),
```
Comment (stephenplusplus) [inline src/partial-result-stream.js]: Removing this step led to a system test meltdown. I've re-worked the way we format/decode values in the latest commit. One test is still failing.

Comment (crwilcox) [reply src/partial-result-stream.js]: the actual code fails also. there are odd values comiing through.
```
code: 'ERR_ASSERTION',
  actual: 'undefinedunt_created_on_     18215',
  expected: 'account_created_on_     18215',
  operator: '==' }
```

Comment (stephenplusplus) [reply src/partial-result-stream.js]: Yikes. Can you send over the test file?

Comment (crwilcox) [reply src/partial-result-stream.js]: https://gist.github.com/crwilcox/fa8bb2fb164afdc29747c489ee34ad8a

**Comment by stephenplusplus** [comment]
Status update:

I took over the PR and did some refactors. Wrote unit tests, all was well. One quick check that the system tests work, then 💥 -- I don't remember the exact number, but let's say ~20 tests failed. I reverted my refactors to see if it was something I did, but unfortunately, same errors.

I realized something we didn't carry over when @crwilcox made his changes was a call to `codec.decode`. Since things were re-arranged from the old way, it wasn't exactly clear where it should go. But, I put it back, and **all but one** test fails. And I can't figure that one out.

Basically, I'm hoping @crwilcox can dive back in for a second to see if he has any ideas. I'll leave some comments on this PR next to the behavioral changes I've made since you last touched the code.

File: src/row-builder.js [hunk 2]
```diff
@@ -72,7 +71,7 @@ RowBuilder.formatValue = function(field, value) {
   }
 
   if (field.code !== 'STRUCT') {
-    return value;
+    return codec.decode(value, field);
```
Comment (stephenplusplus) [inline src/row-builder.js]: Here's where we let codec.decode convert the raw values to our custom values, e.g. `codec.Float`.

File: src/codec.js [hunk 3]
```diff
@@ -118,7 +118,7 @@ codec.generateToJSONFromRow = generateToJSONFromRow;
  * @private
  */
 function decode(value, field) {
-  function decodeValue_(decoded, type) {
+  function decodeValue_(decoded, type, options) {
```
Comment (stephenplusplus) [inline src/codec.js]: TODO: remove `options` (stale code when I was experimenting)

File: src/row-builder.js [hunk 4]
```diff
@@ -46,7 +45,7 @@ RowBuilder.getValue = function(obj) {
   var value = obj;
 
   if (obj && obj.kind) {
-    value = obj[obj.kind];
+    value = commonGrpc.Service.decodeValue_(obj);
```
Comment (stephenplusplus) [inline src/row-builder.js]: This is `decodeValue_`:

```js
GrpcService.decodeValue_ = function(value) {
  switch (value.kind) {
    case 'structValue': {
      return GrpcService.structToObj_(value.structValue);
    }

    case 'nullValue': {
      return null;
    }

    case 'listValue': {
      return value.listValue.values.map(GrpcService.decodeValue_);
    }

    default: {
      return value[value.kind];
    }
  }
};
```

File: src/row-builder.js [hunk 5]
```diff
@@ -150,17 +156,43 @@ RowBuilder.prototype.build = function() {
...
 /**
- * Transforms values into JSON format.
+ * Flush already complete rows.
  */
-RowBuilder.prototype.toJSON = function() {
-  this.build();
+RowBuilder.prototype.flush = function() {
+  var rowsToReturn = this.rows;
 
+  if (!is.empty(this.rows[0]) && this.currentRow.length !== this.fields.length) {
```
Comment (stephenplusplus) [inline src/row-builder.js]: @crwilcox While the previous changes I pointed out were trying to preserve old behavior, this is 1/2 changes to the logic that you wrote:

```diff
-if (!is.empty(this.rows)
+if (!is.empty(this.rows[0])
```

This is because `this.rows` should never be empty, as I understand it. When we reset `this.rows`, we set it to: `[[]]`. So `this.rows` will always have an object, and the is-empty check will always be false.

**Comment by stephenplusplus** [comment]
@crwilcox I think we're finally good again. Units, systems, and your revision of the "issue-180.js" file I've checked are working. It would be great if you could confirm!

**Comment by crwilcox** [comment]
Some system tests are failing locally after the edits:

```
 insert & query
        3) should query in callback mode
        4) should query in promise mode
        5) should query in stream mode
```

Coverage numbers:
image

**Comment by stephenplusplus** [comment]
100% line coverage is what we go for. Can you show more details about the system test errors?

**Comment by crwilcox** [comment]
Sure, but the system tests seem to also be an issue on builds before your commits. 

```
3) Spanner
       Tables
         insert & query
           should query in callback mode:

      Uncaught AssertionError [ERR_ASSERTION]: { SingerId: 'gcloud-tests-id-1278e710',
  Name: 'gcloud-tests-name-1278e711',
  Float: null,
  Int: null,
  Info: null,
  Create deepEqual { SingerId: 'gcloud-tests-id-e12752a0',
  Name: 'gcloud-tests-name-e12752a1',
  Float: 8.2,
  Int: 2,
  Info: <Buffer 67 63 6c 6
      + expected - actual

       {
      -  "Accents": [null]
      -  "Created": [null]
      -  "DOB": [null]
      -  "Float": [null]
      -  "HasGear": [null]
      -  "Info": [null]
      -  "Int": [null]
      -  "Name": "gcloud-tests-name-1278e711"
      -  "PhoneNumbers": [null]
      -  "SingerId": "gcloud-tests-id-1278e710"
      +  "Accents": [
      +    "jamaican"
      +  ]
      +  "Created": [Date: 2018-06-04T16:46:51.082Z]
      +  "DOB": [Date: 1969-08-20T00:00:00.000Z]
      +  "Float": 8.2
      +  "HasGear": true
      +  "Info": [Buffer: [
      +    103
      +    99
      +    108
      +    111
      +    117
      +    100
```

**Comment by stephenplusplus** [comment]
@alexander-fenster ran the system tests and had a different failure. I don't believe these are related to our PR here. @callmehiphop can you run these?

I suppose we could always a merge to let Circle have a go at them, then revert it if necessary.

**Comment by crwilcox** [comment]
Reran the tests and they pass on master and on this branch. I think this is safe to merge.
